### PR TITLE
Treat emoji presentation sequences as fullwidth

### PR DESCRIPTION
**Pull request by Jules-Bertholet**
[UAX11](https://www.unicode.org/reports/tr11/#Recommendations) says:

> [UTS51] *emoji presentation sequences* behave as though they were *East Asian Wide*, regardless of their assigned East_Asian_Width property value.

Lookup is done with a 2-level trie.

### CODE REVIEW

**Comment by Jules-Bertholet** [comment]
In terms of [UTS 51 conformance](https://unicode.org/reports/tr51/#Conformance), with this PR, this crate will give the correct widths for:

- The [basic emoji set](https://unicode.org/reports/tr51/#def_basic_emoji_set), including emoji presentation sequences
- The [emoji keycap sequence set](https://unicode.org/reports/tr51/#def_std_emoji_keycap_sequence_set)
- The [RGI emoji flag sequence set](https://unicode.org/reports/tr51/#def_std_emoji_flag_sequence_set)
- The [RGI emoji tag sequence set](https://unicode.org/reports/tr51/#def_std_emoji_tag_sequence_set)

However, it may overestimate (though never underestimate) the rendered widths of:

- The [RGI emoji modifier sequence set](https://unicode.org/reports/tr51/#def_std_emoji_modifier_sequence_set)
- The [RGI emoji ZWJ sequence set](https://unicode.org/reports/tr51/#def_emoji_ZWJ_sequences)

**Comment by Jules-Bertholet** [comment]
I've replaced the binary search with a better datastructure, and also added a section in the rustdoc on what the exact width rules will be following this PR.

File: scripts/unicode.py [hunk 1]
```diff
@@ -66,12 +68,13 @@ def fetch_open(filename: str):
     """Opens `filename` and return its corresponding file object. If `filename` isn't on disk,
     fetches it from `http://www.unicode.org/Public/UNIDATA/`. Exits with code 1 on failure.
     """
+    basename = os.path.basename(filename)
     if not os.path.exists(os.path.basename(filename)):
```
Comment (Manishearth) [inline scripts/unicode.py:72]: nit: fix this line too

File: src/tables.rs [hunk 2]
```diff
@@ -537,4 +561,101 @@ pub mod charwidth {
         0x56, 0x55, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
         0xFF, 0xFF, 0x5F,
     ];
+
+    /// An array of 256 4-bit nibbles. Index with bytes 9-16 (where LSB is 0)
+    /// of the char you want to test. 0xF means it's not part of a presentation seq,
+    /// anything else means index into the next table.
+    static EMOJI_PRESENTATION_INDEX: [u8; 128] = [
```
Comment (Manishearth) [inline src/tables.rs:568]: suggestion: I'd just generate a function with a match statement for this

File: scripts/unicode.py [hunk 3]
```diff
@@ -388,8 +391,78 @@ def make_tables(
     return tables
 
 
+def load_variation_sequences() -> "list[int]":
+    """Outputs a list of character ranages, corresponding to all the valid characters for starting
+    an emoji presentation sequence."""
+
+    with fetch_open("emoji/emoji-variation-sequences.txt") as sequences:
+        sequence = re.compile(r"^([0-9A-F]+)\s+FE0F\s*;\s+emoji style")
```
Comment (Manishearth) [inline scripts/unicode.py:399]: nit: please document what this regex is doing

**Review by Manishearth — APPROVED** [review]

**Comment by Jules-Bertholet** [comment]
The not-yet-released Unicode 16 adds 8 new non-emoji standardized variation sequences that affect width: https://unicode.org/alloc/Pipeline.html#variation_sequences, https://www.unicode.org/L2/L2023/23212r-quotes-svs-proposal.pdf. In time, we'll need to support those as well.

**Comment by Jules-Bertholet** [comment]
This PR is bloated, I'm splitting it up.

**Pull request closed without merging**